primitives/direct_rpc_server/src/rpc_connection_registry.rs

```diff
@@ -0,0 +1,142 @@
+/*
+	Copyright 2019 Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::sync::SgxRwLock as RwLock;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+use crate::RpcConnectionRegistry;
+use std::collections::HashMap;
+use substratee_worker_primitives::RpcResponse;
+use tls_websocket_server::WebSocketConnection;
+
+type HashMapLock<K, V> = RwLock<HashMap<K, V>>;
+
+pub struct ConnectionRegistry<Hash, Connection>
+where
+	Hash: std::hash::Hash + Eq,
+	Connection: WebSocketConnection,
+{
+	connection_map: HashMapLock<
+		<Self as RpcConnectionRegistry>::Hash,
+		(<Self as RpcConnectionRegistry>::Connection, RpcResponse),
+	>,
+}
```

Comment:
Isn't this cannibalizing itself? A connection registry that itself takes a connection registry as type parameter?

Reply:
According to my understanding, your implementation does require implicitly the same trait bounds. I did not really check if it makes sense in greater context. But your trait bounds seem random because they are not used in the struct itself.

Maybe this makes more sense?

```
pub struct ConnectionRegistry<Hash, Connection>
where
	Hash: std::hash::Hash + Eq,
	Connection: WebSocketConnection,
{
	connection_map: HashMapLock<
		Hash,
		(Connection, RpcResponse),
	>,
}
```

Reply:
Anyhow, I just saw this is outdated code. Maybe my comment is obsolete. :D